Add prefix filter support

zaidoon1 opened the pull request:
rebased https://github.com/fjall-rs/lsm-tree/pull/151 on top of main and adapted various things to the new api

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Prefix extractors for prefix-aware filtering, persisted extractor metadata, and compatibility checks; reads can skip non-matching table ranges to reduce work; writers record/use extractors so filters index prefixes

* **Tests**
  * Extensive unit, integration and recovery tests covering extractor behavior, compatibility, and range-skipping

* **Fuzz**
  * New fuzz harness validating prefix-filter correctness; improved fuzz run setup and ignore rules

* **Benchmarks**
  * Benchmarks measuring range-query performance with and without extractors

* **Chores**
  * Robust directory creation in fuzz run instructions (mkdir -p)
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

marvin-j97 commented:
Unfortunately this has been hit by another wave of conflicts, but I just released 3.0.0, so there will be a bit of a freeze of activity from this point on.

marvin-j97 commented on src/config/mod.rs:
```diff
@@ -308,6 +315,14 @@ impl Config {
         }
     }
 
+    /// Sets the prefix extractor for building prefix-aware Bloom filters.
```
```suggestion
    /// Sets the prefix extractor for building prefix-aware filters.
```

marvin-j97 commented on src/table/writer/mod.rs:
```diff
@@ -92,6 +92,10 @@ pub struct Writer {
 
     initial_level: u8,
+
+    /// Optional prefix extractor used to register extracted prefixes into the Bloom filter.
```
```suggestion
    /// Optional prefix extractor used to register extracted prefixes into the filter.
```

marvin-j97 commented on src/config/mod.rs:
```diff
@@ -226,6 +226,11 @@ pub struct Config {
 
+    /// Optional prefix extractor used to construct prefix-aware filters.
+    /// When set, the table writer will add extracted prefixes (instead of full keys)
+    /// to the Bloom filter and persist the extractor name in table metadata for compatibility checks.
```
```suggestion
    /// to filters and persist the extractor name in table metadata for compatibility checks.
```

marvin-j97 commented on src/table/mod.rs:
```diff
@@ -605,6 +699,67 @@ impl Table {
+            if let Some(mp) = extractor.extract(self.min_key().as_ref()).next() {
+                if sp == mp {
+                    if let Some(sk) = start_key {
+                        if matches!(self.maybe_contains_prefix(sk, extractor), Ok(Some(false))) {
```
... and possibly again.

marvin-j97 commented on src/run_reader.rs:
```diff
@@ -15,46 +18,151 @@ pub struct RunReader {
+    /// indices. It initializes boundary table readers and
+    /// performs lazy per-table prefix-filter skipping during iteration.
     #[must_use]
     pub fn culled<R: RangeBounds<UserKey> + Clone + Send + 'static>(
```
I would recommend running the mutation tests and see if any of these functions pop up in it. Because if there's a single wrong edge case in there it's easy to mess up the read path.
In fact, prefix extractors are so invasive to the read path that it may be worth thinking about fuzz testing them (so we don't get wrongly applied filters or anything).

zaidoon1 replied:
ran the mutations tests, they didnt' find real bugs, but I added more tests to cover the "missed" cases

zaidoon1 replied:
i've added a test file that does fuzz testing so that it runs on each PR. It found a bug and I fixed it as well.

marvin-j97 replied:
Could the fuzz be moved into the fuzz folder with cargo-afl maybe? Then it can be run in an infinite loop, too, and AFL will save all crashes to rerun deterministically.

zaidoon1 committed:
Add AFL fuzz target for prefix filter correctness

Oracle-based differential fuzzer: runs the same AFL-derived operation
sequence against two trees (one with prefix extractor, one without) and
asserts all reads return identical results. Any mismatch = wrongly
applied filter = silent data loss, saved by AFL as a crash for replay.

Covers all identified correctness dimensions:
- 9 extractor variants × 3 bpk levels × 3 filter partitioning policies
- MVCC snapshot reads at older seqnos while writes continue
- Weak tombstones and their compaction GC interaction
- Extractor changes on reopen (prefix_filter_allowed compatibility)
- Partitioned filter forced on all levels (the path that had the panic)
- Bidirectional iterator stepping (PrefixPingPong)
- Unbounded iteration (FirstKV/LastKV)
- Clustered keys (first byte 0..7, len 1..9) for realistic prefix
  distribution with natural in-domain / out-of-domain key mix

zaidoon1 committed:
Annotate defensive dead code and add backward hi-meets-lo test

zaidoon1 committed:
Add prefix filter support

Add prefix-aware filter support to the LSM-tree. When a prefix extractor
is configured, extracted prefixes are added to Bloom filters and the
extractor name is stored in table metadata. Point reads use a prefix
pre-check (maybe_contains_prefix) before falling back to the full-key
filter. Range scans skip tables whose prefix filter definitively excludes
the query range, both upfront and lazily during iteration.

Includes a fix for PartitionedFilterWriter::finish which panicked when
no filter partitions were created (e.g. all keys shorter than the
required prefix length). The empty tli_handles guard returns early
instead of attempting to encode an empty top-level index.

zaidoon1 commented:
i've just pushed a commit, this adds coverage for the last valid case that i missed. The rest is "safe fallbacks" for code that should not be hit. What would prefer? I've added debug_asserts and kept the safe fallback. I can also expect/panic all together if you prefer?

marvin-j97 committed:
refactor: some prefix tests

marvin-j97 committed:
refactor: some prefix tests

marvin-j97 committed:
test: domain prefix extractor doctest

marvin-j97 committed:
doc

marvin-j97 committed:
refactor: run reader prefix test submodule

marvin-j97 committed:
refactor

marvin-j97 committed:
test: empty partitioned filter

marvin-j97 commented:
Found a couple crashes through fuzz testing at 77d0911bc97b143138c5825fd65e05bb18de1b66:

[crashes_1.tar.gz](https://github.com/user-attachments/files/25376014/crashes_1.tar.gz)

zaidoon1 committed:
Replace runtime key-range overlap guards with debug_assert

range_overlap_indexes is exact for disjoint sorted runs — every table
in lo..hi genuinely overlaps the query range. The runtime
check_key_range_overlap guards were provably unreachable dead code.

Replace them with debug_assert to document the invariant and catch
regressions in dev/test builds. Even if the invariant were violated
in production, the impact is benign: table.range() returns an empty
iterator (no data corruption, just wasted I/O).

Remove the three tests that claimed to exercise the guards — they
were valid correctness tests but range_overlap_indexes excluded the
non-overlapping tables before the guards were ever reached.

zaidoon1 commented:
> > i've just pushed a commit, this adds coverage for the last valid case that i missed. The rest is "safe fallbacks" for code that should not be hit. What would prefer? I've added debug_asserts and kept the safe fallback. I can also expect/panic all together if you prefer?
> 
> It's probably fine. The fallback is a bit unfortunate because it just requires an additional range checking - I guess one could just trust the `range_overlap_indexes` function, on the other hand, maybe just removing the safe guard may be fine as well? If `range_overlap_indexes` were to actually return a non-overlapping table, would this be critical?

I've removed the fallback, kept the debug_assert and added comments on what would happen if a refactor broke the code and it wasn't caught by the debug_assert (i.e only wasted I/O)

zaidoon1 committed:
Fix fuzzer: invalidate snapshot seqno after version-changing ops

The fuzzer captures raw snapshot seqnos via vis.get() without pinning
them through the tree's snapshot API. After Compact, MajorCompact, or
Reopen, old super versions may be GC'd, causing
get_version_for_snapshot to panic on the stale seqno.

Reset snapshot_seqno to None after these operations so subsequent
SnapshotGet/SnapshotPrefixScan ops are skipped until a new
TakeSnapshot captures a fresh seqno.

Fixes all 23 AFL crash inputs from the maintainer's fuzz run.

zaidoon1 committed:
Fix data loss: encode prefix length in extractor name

FixedPrefixExtractor::name() returned "fixed_prefix" and
FixedLengthExtractor::name() returned "fixed_length" regardless of
the configured prefix length. After reopening with a different-length
extractor of the same type (e.g. FixedPrefix(3) -> FixedPrefix(4)),
prefix_filter_allowed() saw matching names and allowed the filter to
be probed with the wrong prefix length — producing false negatives
that silently dropped data from point reads, prefix scans, range
scans, and bidirectional iteration.

Now encode the length: "fixed_prefix:3", "fixed_length:4", etc.
This makes prefix_filter_allowed() correctly detect incompatibility
and bypass the filter for tables written with a different length.

Found via AFL fuzz testing (158 crash inputs, all fixed).

marvin-j97 committed:
refactor

marvin-j97 committed:
remove some trait bounds

marvin-j97 committed:
refactor(should_skip_range_by_prefix_filter): dont require owned key

marvin-j97 committed:
add PrefixExtractor::extract_first

marvin-j97 committed:
doc

marvin-j97 committed:
doc

marvin-j97 committed:
doc

zaidoon1 committed:
test: cover partitioned filter spill in register_bytes

zaidoon1 committed:
refactor: replace silent skip with expect in register_bytes spill

marvin-j97 committed:
ignore test coverage for unit test helper fn

marvin-j97 committed:
bump

marvin-j97 committed:
doc: use mkdirp in fuzz test

zaidoon1 committed:
fix: add seqno normalization to get_without_filter and harden fuzz error handling

- get_without_filter now applies the same global_seqno normalization and
  early-exit check as Table::get, preventing potential MVCC visibility
  errors for tables with non-zero global_seqno (defensive fix)
- Replace silent .ok() with .expect() in 7 fuzz call sites so iterator
  errors surface immediately instead of being swallowed
- Add unit test verifying get_without_filter and get agree under
  global_seqno translation (fails without the fix)
- Fix contradictory assertion messages in prefix_filter_recovery tests

zaidoon1 commented on src/table/mod.rs:
```diff
@@ -226,6 +298,23 @@ impl Table {
+        seqno: SeqNo,
+    ) -> crate::Result<Option<InternalValue>> {
+        self.point_read(key, seqno)
+    }
```
this bug is currently unreachable through the public API because:
1. global_seqno != 0 only occurs for ingested tables
2. ingested entries always have internal seqno 0
3. Version-level MVCC filtering (get_version_for_snapshot) ensures the ingested table is only visible when version.seqno < query_seqno, which guarantees global_seqno < query_seqno
4. Under that invariant, both the normalized and unnormalized paths produce identical results

however i've fixed it in case things change in the future

zaidoon1 committed:
Simplify point_read_from_table control flow

Flatten the nested if/else into early returns, eliminating the
intermediate `item` binding. Also fix the trailing comment to
accurately describe both code paths that reach get_without_filter
(prefix filter already consulted, or filter not trustworthy).

zaidoon1 committed:
Adapt test_no_prefix_extractor to upstream filter metrics semantics

Upstream changed filter_queries to only increment on missing key lookups
(issue #246). The test was asserting the old behavior where filter_queries
incremented for every lookup. Use a nonexistent key within the table's
key range to properly exercise the filter counter.

zaidoon1 committed:
Remove unsound multi-prefix range skip and harden metadata parsing

- Remove the second optimization block in should_skip_range_by_prefix_filter
  that could incorrectly skip tables in multi-prefix ranges when the start
  prefix matched the table's min key prefix but the probe landed in a
  different filter partition
- Error on malformed UTF-8 in prefix_extractor metadata instead of silently
  falling back to None
- Document the extract_first() and name() correctness invariants on
  PrefixExtractor
- Return 0 from full filter finish() when no hashes were registered,
  matching partitioned filter behavior

zaidoon1 committed:
Remove duplicate prefix_extractor_name field from Inner

Read through metadata.prefix_extractor_name instead of maintaining a
separate copy on Inner, eliminating the clone and the invariant that
both fields must stay in sync.

zaidoon1 committed:
Simplify PrefixExtractor doc example and fix typo

Use the exported FixedPrefixExtractor in the doc example instead of
reimplementing it inline. Change 'iff' to 'if' in extract_first docs.

zaidoon1 commented on src/prefix.rs:
```diff
@@ -0,0 +1,364 @@
+    /// however it can be overridden to skip the Box allocation of `extract` in some cases.
+    ///
+    /// Implementations that override this method must remain semantically identical to
+    /// `self.extract(key).next()`: return `None` iff `extract` would yield no prefixes,
```
fixed

zaidoon1 committed:
Fix prefix filter correctness bugs and optimize RunReader prefix pruning

Three correctness bugs fixed:

1. FixedPrefixExtractor false negatives for short keys: extract_first()
   returned Some(short_key) for keys shorter than the configured length,
   producing a hash that was never in the filter. Changed to return None
   (out-of-domain), matching RocksDB's NewFixedPrefixTransform behavior.
   This caused silent data loss when tree.prefix() was called with a
   prefix shorter than the extractor length.

2. tree.prefix() could not use the filter when the prefix length equaled
   the extractor length: prefix_to_range("h") produces range "h".."i",
   and extract_first("h") != extract_first("i"), so the filter layer
   treated it as a multi-prefix range and never consulted the filter.
   Fixed by threading the original prefix from tree.prefix() through to
   the filter layer as a prefix_hint, bypassing the range-based prefix
   extraction. A stability guard (extract_first(hint) == extract_first(
   hint + "\0")) prevents false negatives when the hint is not a valid
   extracted prefix (e.g. FullKeyExtractor, or hint shorter than extractor
   length).

3. RunReader lazy per-table skip missing prefix_filter_allowed check:
   the optimized validated_prefix_hint path called probe_prefix_filter
   directly without verifying extractor compatibility. Since optimize_runs
   can merge tables from different compaction epochs (with different
   extractor configs) into a single run, this could probe an old table's
   filter with the wrong extractor and get a false negative.

Performance optimizations in RunReader:

- Precompute the validated prefix hint once in RunReader::new instead of
  re-running the stability guard (with a Vec allocation) on every table
  during lazy iteration
- Replace common_prefix: Option<Vec<u8>> with can_prune_upfront: bool,
  eliminating up to 3 .to_vec() allocations per RunReader::new
- Use extractor.as_ref() instead of extractor.clone() in the upfront
  pruning loop, avoiding an unnecessary Arc refcount bump
- Call probe_prefix_filter directly in the lazy loops when a validated
  hint is available, skipping the full should_skip_range_by_prefix_filter
  guard re-check

zaidoon1 committed:
Fix clippy: collapse nested if, move const before statements, take prefix_hint by reference

zaidoon1 committed:
Improve fuzz target coverage for prefix filter edge cases

- Expand ClusteredPrefix range from 0-3 to 0-5 bytes so prefix length
  can equal or exceed all extractor lengths (up to 4), covering the
  case where prefix_to_range produces bounds with different extracted
  prefixes
- Widen prefix byte alphabet from 0-7 to 0-9 so values 8-9 produce
  prefixes absent from all keys, exercising the filter-should-skip path
- Add AFL-controlled MajorCompact target size (128 vs 4096 bytes) to
  produce many small tables per run, making the RunReader lazy loop
  (3+ table overlap) far more likely to fire
- Add PrefixScanExistingKey operation that derives the scan prefix from
  a previously inserted key, guaranteeing the prefix overlaps real data
  and forcing the filter to make a meaningful decision
- Add FlushCompactReopenCompact composite operation that atomically
  creates the mixed-extractor multi-table run condition (flush, compact
  with small target, reopen with different extractor) without relying
  on AFL to randomly sequence four separate operations

zaidoon1 commented on src/table/writer/filter/full.rs:
I took a look and RocksDB deduplicates adjacent hashes. Since keys with the same prefix are sorted adjacently in an SST file, this effectively deduplicates all prefix hashes within a data block / partition. For the example you gave: `<key>#<version>` with 10 versions per key, all 10 versions are adjacent, so only 1 hash is added.

zaidoon1 replied:
that's correct. and Rocksdb doesn't have have to worry about it because it doesn't support multi prefix extractors. I'll work on pushing a commit to do the deduping correctly

zaidoon1 committed:
Deduplicate prefix hashes in filter writers to avoid oversized filters

When a prefix extractor is configured, multiple keys sharing the same
prefix produce duplicate hashes in the Bloom filter buffer. Without
dedup, the filter is sized for N total hashes instead of the true number
of unique prefixes, wasting memory proportional to the average number of
keys per prefix.

Sort + dedup the hash buffer at flush/spill time so the filter is sized
for the actual number of unique prefixes. This is gated behind an
enable_dedup() flag set by Writer::use_prefix_extractor, so the full-key
Bloom path (where each hash is already unique) pays no overhead.

For partitioned filters, dedup runs per-partition in spill_filter_partition
before building each partition's Bloom filter.

Also add a HierarchicalExtractor to the fuzz target that returns multiple
prefixes per key (2-byte + 4-byte), exercising the interleaved hash
dedup path and the multi-prefix contains_prefix probe logic.

marvin-j97 commented on src/table/mod.rs:
```diff
@@ -310,6 +373,49 @@ impl Table {
+            {
+                use std::sync::atomic::Ordering::Relaxed;
+                self.metrics.filter_queries.fetch_add(1, Relaxed);
+            }
```
I think this negatively affects filter metrics - because the filter queries are incremented, but on hit there is no increment for filter hits.

zaidoon1 replied:
pushed a commit to fix this, let me know what you think

marvin-j97 commented:
I have run a couple of benchmarks with varying number of items and duration, and I've been getting mixed results so far. Even when artificially making the workload IO-bound, the performance improvements were often <10%.
In the column store workload (with a fixed prefix of 48), the performance actually degraded by 10%, but I'm not yet really sure why.

zaidoon1 commented:
> I have run a couple of benchmarks with varying number of items and duration, and I've been getting mixed results so far. Even when artificially making the workload IO-bound, the performance improvements were often <10%. In the column store workload (with a fixed prefix of 48), the performance actually degraded by 10%, but I'm not yet really sure why.

can you give me the commands you are running so i can try to reproduce this locally and dig deeper?

marvin-j97 commented:
> can you give me the commands you are running so i can try to reproduce this locally and dig deeper?

Feed:
```nu
let data_dir = ".data"
let seconds = 10 * 60
let cache_bytes = 256_000_000

alias bench = cargo run -r --features mimalloc,metrics --

bench run --cache-size $cache_bytes --auto-granularity 80 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "pfx_.jsonl" --display-name "fjall 3.1.2" --backend fjall3 feed --users 5000000 --zipf-exponent 0.5
bench run --cache-size $cache_bytes --auto-granularity 80 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "pfx_with.jsonl" --display-name "fjall 3.1.2 pfx" --lsm-prefix-extractor --color "#eb5aaaff" --backend fjall3 feed --users 5000000 --zipf-exponent 0.5
bench report pfx_*.jsonl --out report_pfx.html
google-chrome-stable report_pfx.html
```

Column Store:
```nu
let data_dir = ".data";
let seconds = 20 * 60;
let cache_bytes = 1_000_000_000;
let value_size = 100;
let db_count = 1_000;
let threads = 4;

alias bench = cargo run -r --features mimalloc,metrics --

bench run --auto-granularity 100 --cache-size $cache_bytes --lsm-block-size 16000 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "cs_pfx_.jsonl" --display-name "fjall 3.1.2" --backend fjall3 column-store --value-size $value_size --readers $threads --database-count $db_count
RUST_LOG=warn bench run --auto-granularity 100 --cache-size $cache_bytes --lsm-block-size 16000 --data-dir $data_dir --seconds $seconds --out "cs_pfx_with.jsonl" --display-name "fjall 3.1.2 pfx" --lsm-prefix-extractor --color "#eb5aaaff" --backend fjall3 column-store --value-size $value_size --readers $threads --database-count $db_count
bench report cs_pfx_*.jsonl --out report_cs_pfx.html
google-chrome-stable report_cs_pfx.html
```

Note that I replaced `--lsm-prefix-extractor-len` with a workload-dependent extractor.

zaidoon1 committed:
Move prefix filter metrics to callers for accurate per-context tracking

Remove metrics from probe_prefix_filter entirely and let each caller
track filter_queries and io_skipped_by_filter in a context-appropriate
way, matching the upstream Table::get() pattern from issue #246.

Point reads (point_read_from_table):
- Filter excludes: both counters incremented (saved I/O)
- Filter allows, key not found: filter_queries only (wasted I/O)
- Filter allows, key found: no increment (successful read)

Range scans (should_skip_range_by_prefix_filter, RunReader lazy loops):
- Filter excludes: both counters incremented (saved I/O)
- Filter allows: no increment (outcome unknown until iteration)

This ensures io_skipped_by_filter / filter_queries gives the true
positive rate of filter decisions, without inflating filter_queries
for successful reads.

marvin-j97 commented on src/table/mod.rs:
```diff
@@ -623,6 +727,92 @@ impl Table {
+        let end_key = match range.end_bound() {
+            std::ops::Bound::Included(k) | std::ops::Bound::Excluded(k) => Some(k.as_ref()),
+            std::ops::Bound::Unbounded => None,
+        };
+
+        let start_pref = start_key.and_then(|k| extractor.extract_first(k));
+        let end_pref = end_key.and_then(|k| extractor.extract_first(k));
+
+        if let (Some(sp), Some(ep)) = (start_pref, end_pref) {
+            if sp == ep {
+                if let Some(sk) = start_key {
+                    if matches!(self.probe_prefix_filter(sk, extractor), Ok(Some(false))) {
+                        #[cfg(feature = "metrics")]
+                        {
+                            use std::sync::atomic::Ordering::Relaxed;
+                            self.metrics.filter_queries.fetch_add(1, Relaxed);
+                            self.metrics.io_skipped_by_filter.fetch_add(1, Relaxed);
+                        }
+
+                        return true;
+                    }
```
Need to increment the filter queries even if there is no true positive.
Otherwise the FPR will be too low.
(And everywhere else)

```suggestion
                if let Some(sk) = start_key {
                    #[cfg(feature = "metrics")]
                    self.metrics.filter_queries.fetch_add(1, Relaxed);

                    if matches!(self.probe_prefix_filter(sk, extractor), Ok(Some(false))) {
                        #[cfg(feature = "metrics")]
                        {
                            use std::sync::atomic::Ordering::Relaxed;

                            self.metrics.io_skipped_by_filter.fetch_add(1, Relaxed);
                        }

                        return true;
                    }
```

zaidoon1 replied:
updated

zaidoon1 commented:
> > can you give me the commands you are running so i can try to reproduce this locally and dig deeper?
> 
> Feed:
> 
> ```nu
> let data_dir = ".data"
> let seconds = 10 * 60
> let cache_bytes = 256_000_000
> 
> alias bench = cargo run -r --features mimalloc,metrics --
> 
> bench run --cache-size $cache_bytes --auto-granularity 80 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "pfx_.jsonl" --display-name "fjall 3.1.2" --backend fjall3 feed --users 5000000 --zipf-exponent 0.5
> bench run --cache-size $cache_bytes --auto-granularity 80 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "pfx_with.jsonl" --display-name "fjall 3.1.2 pfx" --lsm-prefix-extractor --color "#eb5aaaff" --backend fjall3 feed --users 5000000 --zipf-exponent 0.5
> bench report pfx_*.jsonl --out report_pfx.html
> google-chrome-stable report_pfx.html
> ```
> 
> Column Store:
> 
> ```nu
> let data_dir = ".data";
> let seconds = 20 * 60;
> let cache_bytes = 1_000_000_000;
> let value_size = 100;
> let db_count = 1_000;
> let threads = 4;
> 
> alias bench = cargo run -r --features mimalloc,metrics --
> 
> bench run --auto-granularity 100 --cache-size $cache_bytes --lsm-block-size 16000 --lsm-use-partitioned-meta --data-dir $data_dir --seconds $seconds --out "cs_pfx_.jsonl" --display-name "fjall 3.1.2" --backend fjall3 column-store --value-size $value_size --readers $threads --database-count $db_count
> RUST_LOG=warn bench run --auto-granularity 100 --cache-size $cache_bytes --lsm-block-size 16000 --data-dir $data_dir --seconds $seconds --out "cs_pfx_with.jsonl" --display-name "fjall 3.1.2 pfx" --lsm-prefix-extractor --color "#eb5aaaff" --backend fjall3 column-store --value-size $value_size --readers $threads --database-count $db_count
> bench report cs_pfx_*.jsonl --out report_cs_pfx.html
> google-chrome-stable report_cs_pfx.html
> ```
> 
> Note that I replaced `--lsm-prefix-extractor-len` with a workload-dependent extractor.

I believe i found the issue. When a prefix extractor was configured, the filter only contained prefix hashes, full-key hashes were not written. For point reads where the key doesn't exist but its prefix DOES exist in the table (e.g., looking up user_alice when user_bob exists — same prefix "user"):
- Without prefix extractor: full-key Bloom catches the nonexistent key ~99% of the time → no data block read
- With prefix extractor: prefix filter says "maybe present" → get_without_filter → reads data blocks → finds nothing → wasted I/O

Rocksdb solves it using the whole_key_filtering option that is default enabled which adds both the prefix + the whole key to the bloom unless you disable whole_key_filtering in which case only the prefix is added: https://github.com/facebook/rocksdb/wiki/RocksDB-Bloom-Filter#prefix-vs-whole-key

I'll add the same thing here and make it configurable because I have workloads in prod that disables whole_key_filtering because I don't do point lookups so I avoid the wasted space.

marvin-j97 commented:
> I believe i found the issue. When a prefix extractor was configured, the filter only contained prefix hashes, full-key hashes were not written. For point reads where the key doesn't exist but its prefix DOES exist in the table (e.g., looking up user_alice when user_bob exists — same prefix "user"):
> 
> * Without prefix extractor: full-key Bloom catches the nonexistent key ~99% of the time → no data block read
> * With prefix extractor: prefix filter says "maybe present" → get_without_filter → reads data blocks → finds nothing → wasted I/O
> 
> Rocksdb solves it using the whole_key_filtering option that is default enabled which adds both the prefix + the whole key to the bloom unless you disable whole_key_filtering in which case only the prefix is added: https://github.com/facebook/rocksdb/wiki/RocksDB-Bloom-Filter#prefix-vs-whole-key
> 
> I'll add the same thing here and make it configurable because I have workloads in prod that disables whole_key_filtering because I don't do point lookups so I avoid the wasted space.

Yeah, I'm aware of full keys. I have actually manually extracted the full keys in the custom extractor for the `Feed` workload. That should take care of it unless the extracted full key is not actually used (haven't checked tbh). However, the column store workload does not even do any point reads.

Edit: Actually manually extracting the full keys may not work so that may at least explain why I saw the feed workload being slower in point reads sometimes.

zaidoon1 committed:
Register both full-key and prefix hashes in the filter (whole_key_filtering)

When a prefix extractor is configured, the filter previously only
contained prefix hashes, causing point reads to bypass the Bloom filter
entirely (via get_without_filter) after the prefix check passed. For
keys that don't exist but share a prefix with existing keys, this meant
reading data blocks only to find nothing — wasted I/O that a full-key
Bloom would have prevented.

Now the filter contains both prefix hashes AND full-key hashes (matching
RocksDB's whole_key_filtering + prefix_extractor approach). Point reads
use two-level filtering: the prefix filter for coarse table-level
pruning, then the full-key Bloom for precise per-key filtering. This
eliminates the 10% regression observed in workloads with many keys per
prefix.

The whole_key_filtering option (default true) can be set to false for
seek-only workloads that never perform point lookups, saving filter
space by omitting full-key hashes.

zaidoon1 commented:
> However, the column store workload does not even do any point reads.

I'll look into this more.

marvin-j97 commented:
So with the whole key filtering at least point reads are the same as before. The workload I've tested is not very write intensive so L0 only gets up to 4 tables shortly, so I guess for a write intensive workload, or setting L0 threshold to be higher, prefix filtering would help more.

zaidoon1 committed:
Fix prefix filter metrics to count every real probe for accurate FPR

Increment filter_queries on every probe where the filter actually exists
and answers (Ok(Some(_))), not just on skips. This ensures
io_skipped_by_filter / filter_queries gives the true positive rate of
filter decisions. Previously filter_queries only incremented on definitive
exclusion, making the FPR appear artificially low.

Return Ok(None) from probe_prefix_filter and maybe_contains_prefix when
the filter cannot answer (no filter block, incompatible extractor, or
out-of-domain key). Callers only increment metrics for Ok(Some(_)),
preventing false counts for tables without filters or with mismatched
extractors.

All 5 prefix filter call sites (point_read_from_table, both paths in
should_skip_range_by_prefix_filter, and both RunReader lazy loops) now
follow the same pattern: capture the probe result, count if it answered,
count the skip if it excluded.

zaidoon1 committed:
Precompute prefix hash in RunReader to eliminate per-table allocation

The RunReader lazy loop previously called probe_prefix_filter on every
table, which internally called extractor.extract(key) (a Box<dyn Iterator>
heap allocation) and re-hashed the prefix bytes on each probe. For
workloads scanning hundreds of tables per query, this overhead was the
primary cause of a 10.6% throughput regression in the column store
benchmark.

Precompute the prefix hash once in RunReader::new (from the validated
prefix hint) and store it alongside the hint. The lazy loops now call
probe_prefix_filter_with_hash which uses the key only for TLI partition
selection and checks the precomputed hash directly against the Bloom
filter, eliminating:
- Box<dyn Iterator> allocation per table
- extract_first() call per table
- Hash computation per table

Column store benchmark results (5 min, 1000 databases, 48-byte prefix):
- Before optimization: -10.6% regression vs baseline
- After optimization: +16.1% improvement vs baseline
- Tail latency: 966us -> 27us (36x improvement)

zaidoon1 commented:
I've pushed up a commit to fix the regression:

```
=== COLUMN STORE BENCHMARK COMPARISON (5 min each) ===
Config                          Scans/s  vs base    Lat(us)  Filter(MiB)  DiskRd(MiB)
-------------------------------------------------------------------------------------
Baseline (no prefix)             32,155    +0.0%       32.3        304.8     64,062.4
Prefix 48B (pre-opt)             28,755   -10.6%      966.9        609.7     52,248.1
Prefix 48B (optimized)           37,330   +16.1%       27.2        609.4     53,634.5
Prefix 48B (no wkf)              35,452   +10.3%       28.8        304.6     66,498.8
```

zaidoon1 committed:
Use extract_first instead of extract in contains_prefix to avoid Box allocation

contains_prefix previously called extractor.extract(key) which returns
Box<dyn Iterator> — a heap allocation on every filter probe. Replace
with extract_first(key) which returns Option<&[u8]> (zero allocation).

Checking only the first prefix is sufficient for filter probing: during
writes, extract_first(key) is always registered in the filter for every
key. If it is absent, the key was never written and the table can be
safely skipped. Secondary prefixes from extract() can only match hashes
from other keys with different first prefixes, which is irrelevant.

Feed workload benchmark (point-read heavy, 5 min):
- Point read throughput: +18.1%
- Point read latency: -30.4%
- Range scan throughput: +12.8%

zaidoon1 commented:
and the extract vs extract_first optimization:
```
=== FEED WORKLOAD: extract() vs extract_first() (5 min each) ===
Metric                      Box (extract)   extract_first     Change
----------------------------------------------------------------------
Point read ops/s                  245,881         290,299     +18.1%
Range scan ops/s                   63,874          72,052     +12.8%
Write ops/s                       235,287         278,494     +18.4%
Point read lat (us)                   5.0             3.5     -30.4%
Range scan lat (us)                  19.0            14.2     -25.1%
Total point reads              14,846,503      16,609,555     +11.9%
…
```

zaidoon1 commented:
perf optimizations wise. I'm not seeing anything else to optimize so I think I can stop here. I'll wait for you to review the latests commits and then address any feedback you may have before i fix the merge conflict.

marvin-j97 commented:
I haven't thought about it too much until now, but at this point only `extract_first` is really used, and makes me question how well multi prefixes really work. Because if you extract a multiple prefixes from a key, let's say `<company_id>#<user_id>#...` -> `["<company>#", "<company>#<user>#"]` naturally that allows to filter for prefix queries over `company` and `(company + user)`.

However: as it currently stands, for a `(company + user)` query, extract_first will only check for `company#`, and I'm unsure whether going through all prefixes warrants the heap allocation + filter lookup overhead.
Obviously `company + user` has a higher cardinality, so it would be more exact for filtering, I guess.

zaidoon1 committed:
Use most specific prefix hash for RunReader filter pruning

Add extract_last to PrefixExtractor trait, returning the highest-
cardinality prefix instead of the coarsest. For multi-prefix extractors
like [company#, company#user#], this allows RunReader to check
hash(company#user#) instead of hash(company#), skipping tables that
contain the company but not the specific user.

The approach naturally adapts to hint length: long hints get the most
specific prefix, short hints fall back to the coarsest. A stability
guard (extract_last(hint) == extract_last(hint + NUL)) ensures the
chosen prefix is stable across the query range. If the guard fails,
falls back to extract_first (previous behavior).

For single-prefix extractors, extract_last is overridden to delegate to
extract_first (zero allocation, identical behavior). The struct size
(Option<u64>) and per-table cost (single Bloom probe) are unchanged.
The only added cost is one extract_last call in RunReader::new (once per
query, not per table).

zaidoon1 committed:
Use extract_last in single-table prefix filter path for consistent pruning

The single-table skip path (should_skip_range_by_prefix_filter with a
prefix hint) now uses the same extract_last optimization as RunReader:
precompute the most specific stable prefix hash, then probe with
probe_prefix_filter_with_hash. Previously it used extract_first, giving
coarser pruning than the RunReader path for multi-prefix extractors.

Both prefix filter pruning paths (RunReader lazy loop and single-table
skip) now consistently use the most specific available prefix hash,
with extract_last stability guard and extract_first fallback.

zaidoon1 committed:
Skip redundant prefix pre-check for point reads when whole_key_filtering is enabled

When whole_key_filtering is true (the default), the filter contains
full-key hashes. For point reads, the full-key Bloom is strictly more
precise than the prefix pre-check — so skip the prefix probe entirely
and go straight to the Bloom. This eliminates a redundant filter probe
on every point read that was costing ~10% throughput on point-read-heavy
workloads.

The prefix filter now only activates for range/prefix scans, where it
provides table-level pruning that the full-key Bloom cannot.

When whole_key_filtering is false, point reads still use the prefix
filter as the sole pre-check since no full-key hashes are available.

Feed workload (point-read heavy):
- Before: -9.5% regression vs no prefix extractor
- After: +7.7% improvement vs no prefix extractor, -69.6% disk reads

Column store (scan-only): +4.7% improvement, consistent with previous

zaidoon1 commented:
> I haven't thought about it too much until now, but at this point only `extract_first` is really used, and makes me question how well multi prefixes really work. Because if you extract a multiple prefixes from a key, let's say `<company_id>#<user_id>#...` -> `["<company>#", "<company>#<user>#"]` naturally that allows to filter for prefix queries over `company` and `(company + user)`.
> 
> However: as it currently stands, for a `(company + user)` query, extract_first will only check for `company#`, and I'm unsure whether going through all prefixes warrants the heap allocation + filter lookup overhead. Obviously `company + user` has a higher cardinality, so it would be more exact for filtering, I guess.

ok i've pushed three commits to address all of this:

The `extract()` iterator is still used on the write path to register ALL prefix hashes in the filter, so multi-prefix extractors work correctly for both coarse and fine-grained queries.
For reads, we now use `extract_last` (the most specific prefix) instead of `extract_first` (the coarsest) when probing the filter. This is computed once per query in `RunReader::new` and `should_skip_range_by_prefix_filter`, not per table — so the one-time `extract()` Box allocation is amortized. The per-table probe uses a precomputed hash with zero allocation.
For a multi-prefix extractor like `["company#", "company#user#"]`:
- `tree.prefix("company#user#hello")` → probes with `hash("company#user#")` (most specific stable prefix). Tables with the company but not this specific user are skipped.
- `tree.prefix("company#")` → probes with `hash("company#")` (only prefix available at this hint length). Tables without this company are skipped.

The approach adapts to the hint length automatically. A stability guard (`extract_last(hint) == extract_last(hint + "\0")`) ensures the chosen prefix is stable across the query range — if the most specific prefix isn't stable, it falls back to `extract_first`.
For single-prefix extractors, `extract_last` is overridden to delegate to `extract_first` (zero allocation, identical behavior). No overhead change.
For point reads with `whole_key_filtering=true` (the default), the prefix filter is skipped entirely — point reads go straight to the full-key Bloom which is strictly more precise. The prefix filter only activates for range/prefix scans where it provides table-level pruning.

zaidoon1 committed:
Persist whole_key_filtering in table metadata

Tables built with a prefix extractor now record their whole_key_filtering
value in the metadata block. Reads use the table's persisted value to
choose between the full-key Bloom and the prefix-only path, so reopening
a tree with a different runtime configuration cannot route reads through
a filter that was never populated for that table.

Tables written before this change (which always populated full-key
hashes alongside prefix hashes) default to whole_key_filtering=true on
recovery for backward compatibility.

zaidoon1 committed:
Spill partitioned filter only on user-key boundaries

The partitioned filter writer used to check the spill threshold inside
register_bytes. When a key's hashes were registered as multiple separate
calls (prefix hashes + full-key hash), a spill could occur partway
through the key. The spilled partition's TLI key corresponds to the
current user key, but only the hashes registered before the spill end
up in that partition; the rest move to the next partition. A subsequent
filter probe for that user key follows the TLI to the wrong partition
and misses the deferred hashes, causing false negatives.

Defer the spill check to user-key boundaries: on the next notify_key
(spilling using the previous last_key) and after register_key (whose
hash is the last contribution for its user key). register_bytes only
buffers hashes; it no longer spills. Buffered hashes for a single key
are guaranteed to land together in one partition, so a partition's TLI
key always maps to fully committed hashes for that key.

zaidoon1 committed:
Update prefix filter doc comments to match current behavior

Several code comments described old behavior that has since changed:

- Writer's prefix-then-full-key ordering rationale referred to
  register_bytes triggering partition spills, but register_bytes no
  longer spills after the recent partition-boundary fix.
- should_skip_range_by_prefix_filter and RunReader::validated_prefix_hint
  comments only described extract_first stability; both code paths now
  prefer extract_last with extract_first as a fallback.
- StandardBloomFilterReader::contains_prefix and FilterBlock and
  Table::maybe_contains_prefix all said 'any extracted prefix' but only
  consult extract_first.
- Table::maybe_contains_prefix doc claimed Ok(Some(true)) when no filter
  is available; actual return is Ok(None).
- FilterWriter::notify_key trait doc lacked an Errors section after the
  signature changed to return Result.
- PartitionedFilterWriter::notify_key inline comment lumped
  register_bytes and register_key together as 'mid-key', conflicting
  with register_key's own comment that documents it as a safe spill
  point at the end of a user key.
- The reverse iteration path in RunReader was missing the
  prefix_filter_allowed-per-table rationale that the forward path has.

zaidoon1 committed:
Fix clippy warnings: is_none_or, if_not_else, must_use, needless borrows

zaidoon1 committed:
Propagate dedup flag when use_partitioned_filter replaces filter writer

Writer::use_prefix_extractor calls enable_dedup() on the *current*
filter writer, but use_partitioned_filter() *replaces* it with a fresh
PartitionedFilterWriter (needs_dedup=false). When callers invoke
use_prefix_extractor before use_partitioned_filter — which is the
order in tree flush, bulk ingest, blob-tree flush, and MultiWriter
rotation — the dedup flag is silently lost on the discarded writer.

Result: every duplicate prefix hash gets re-buffered, inflating
bloom_hash_buffer.len() (used as n for filter sizing) and triggering
premature partition spills. Functional correctness is preserved
because Bloom filters are set-bit-idempotent, but partitioned filters
are over-sized and over-spilled in proportion to the average number
of keys per prefix.

Re-apply enable_dedup() inside use_partitioned_filter when a prefix
extractor is configured. Add regression tests covering both the flush
path and the MultiWriter::rotate path (forced via a small target_size
on major_compact).

zaidoon1 committed:
Use full-key Bloom on extractor mismatch when whole_key_filtering=true

When point_read_from_table cannot trust the prefix portion of a
table's filter (extractor names mismatch, current config has no
extractor, or the table was written before extractors existed), it
previously fell straight through to a data-block scan via
get_without_filter — discarding the full-key Bloom even when it
remained valid.

The full-key Bloom is independent of the prefix extractor: tables
written with whole_key_filtering=true contain full-key hashes
regardless of which extractor (or none) was active at write time.
Legacy tables predate extractor support entirely and always wrote
full-key hashes, parsing whole_key_filtering=true by default.

Consult the full-key Bloom in this case before falling through.
Tables written with whole_key_filtering=false and a now-mismatched
extractor still skip the filter — their filter only holds prefix
hashes that aren't queryable with the full-key hash, so the
get_without_filter path is required to avoid false negatives.

Add tests covering all three mismatch paths (different extractor,
legacy table reopened with extractor, extractor dropped on reopen)
plus a negative test confirming the WKF=false case correctly skips
the filter and reads the data block directly.

zaidoon1 commented:
Ran rust-storage-bench (Feed and Column-Store) comparing the basic SST-level prefix Bloom in this PR against RocksDB. The bench harness only calls `set_prefix_extractor` on RocksDB — no memtable bloom, no hash data block index, no other RocksDB-specific features — so this isolates the basic SST-level prefix Bloom comparison.

Same flags everywhere: `--lsm-use-partitioned-meta` on every config, 5-minute measurement window after pre-ingest, fresh data dir per run. M4 Pro / 48 GiB RAM, lsm-tree branch at `9cf39ad9`.

## Feed (5M users, 255M items, 11-byte user prefix)

| | write p50 | write p99 | point_read p50 | point_read p99 | range_read p50 | range_read p99 |
|---|---:|---:|---:|---:|---:|---:|
| lsm-tree baseline | 4.68 µs | 14.92 µs | 99.74 µs | 255.34 µs | 8.35 µs | 121.82 µs |
| lsm-tree +pfx-11  | **4.68 µs** | **14.92 µs** | **95.83 µs** | **213.28 µs** | **7.56 µs** | **117.05 µs** |
| rocksdb  baseline | 5.27 µs | 16.82 µs | 110.23 µs | 255.34 µs | 10.20 µs | 137.36 µs |
| rocksdb  +pfx-11  | 5.38 µs | 17.51 µs | 114.73 µs | 255.34 µs | 10.62 µs | 142.96 µs |

lsm-tree (with or without the extractor) is faster than RocksDB on every metric. Comparing best-vs-best (both with prefix), lsm-tree wins by 13–29% across the board. Enabling the prefix extractor on lsm-tree gives a real benefit: point_read p99 −16.5%, range_read p50 −9.5%.

## Column-Store (1000 DBs, 4 readers, 100B values, 48-byte prefix)

| | write p50 | write p99 | range_read p50 | range_read p90 | range_read p99 |
|---|---:|---:|---:|---:|---:|
| lsm-tree baseline | 3.61 µs | 33.87 µs | 18.22 µs | 38.19 µs | 217.59 µs |
| lsm-tree +pfx-48  | 3.20 µs | **10.20 µs** | **17.86 µs** | **51.55 µs** | 240.47 µs |
| rocksdb  baseline | **2.95 µs** | 8.69 µs | 18.96 µs | 88.46 µs | 282.20 µs |
| rocksdb  +pfx-48  | **2.89 µs** | **8.02 µs** | 18.59 µs | 75.38 µs | **240.47 µs** |

Mixed here. Range reads: lsm-tree wins p50 by 3.9% and p90 by 31.6%, p99 ties. Writes: RocksDB wins p50 by 9.7% and p99 by 21.3%.

The column-store write gap isn't a prefix-filter issue — even RocksDB baseline (no extractor) beats lsm-tree baseline on writes (p99 8.69 vs 33.87 µs) under this workload. That's a write-throttling / compaction-backpressure characteristic, not the filter code path. Adding the prefix extractor narrows lsm-tree's write tail substantially (33.87 → 10.20 µs, since smaller filters reduce compaction backlog) but doesn't close the gap entirely.

Both backends benefit from enabling the prefix extractor on column-store: RocksDB range_read p99 −14.8%, lsm-tree range_read p50 −2%.

Caveat: the rocksdb column-store runs hit `Corruption: Mismatch in unique ID on table file` mid-run — a known rust-rocksdb 0.44.2 issue on macOS (the `Actual` value `{6918025063187695999, 149668763839083335}` is a constant across runs, suggesting an uninitialised-memory read in `verify_sst_unique_id_in_manifest`). The bench produced final histograms before crashing, and the values reproduced consistently across two attempts, so I'm using them with the caveat noted. Linux runs don't hit this.

## Commands

```bash
# Feed
$BENCH run --cache-size 256000000 --auto-granularity 80 --lsm-use-partitioned-meta \
  --data-dir /tmp/feed_data --seconds 300 --out feed_lsm_baseline.jsonl \
  --display-name "lsm-tree baseline" --backend fjall3 \
  feed --users 5000000 --zipf-exponent 0.5

# repeat with --lsm-prefix-extractor-len 11 for the +pfx variant
# repeat with --backend rocksdb for the rocksdb variants

# Column-Store
$BENCH run --auto-granularity 100 --cache-size 1000000000 --lsm-block-size 16000 \
  --lsm-use-partitioned-meta --data-dir /tmp/cs_data --seconds 300 \
  --out cs_lsm_baseline.jsonl --display-name "lsm-tree cs baseline" --backend fjall3 \
  column-store --value-size 100 --readers 4 --database-count 1000

# repeat with --lsm-prefix-extractor-len 48 for +pfx
# repeat with --backend rocksdb for the rocksdb variants
```

Built with `cargo build -r --features mimalloc,rocksdb`. The `metrics` feature is broken in this fork (gates rocksdb code unconditionally), so left it off. Cleared the data dir between runs.

## Summary

Basic SST-level prefix Bloom in this PR is at parity with or ahead of RocksDB's. lsm-tree wins decisively on Feed across all metrics. On Column-Store, lsm-tree wins range reads; RocksDB wins writes — but the write difference is a write-throttling characteristic visible in the baseline too, not the prefix filter.

zaidoon1 committed:
Prefix filter: optimize hot paths, document contracts, expand tests

Hot path:
- Writer skips Box<dyn Iterator> alloc for single-prefix extractors via
  a new is_single_prefix() trait method (default false; built-ins
  override to true).
- TreeIter precomputes the validated prefix hash once at iter setup;
  L0 single-table loop and RunReader upfront pruning both reuse it
  instead of re-extracting and re-hashing per table.
- Hoist prefix_extractor.name() out of the per-table point-read loop.
- Materialize range bounds once outside the per-run loop.

Refactor:
- Extract validate_hint_and_hash helper (shared stability guard logic
  that was duplicated across RunReader::new and
  Table::should_skip_range_by_prefix_filter).
- point_read_from_table now dispatches via a PointReadRoute enum
  (FullKeyBloom / PrefixOnly / NoFilter).
- Writer protocol: notify_key() is now called for every user key in
  every mode; register_key drops the redundant last_key write. Added
  an assert! to enforce the protocol.

Docs:
- PrefixExtractor Contract section: method consistency, name() stability,
  reserved namespaces (fixed_prefix:/fixed_length:/full_key), panic
  safety. User-impl examples moved from the type alias to the trait.
- FilterWriter trait Contract: per-key call order + per-method invariants.
- Config::whole_key_filtering: no-op-without-extractor case documented
  prominently.
- RunReader / IterState struct docs updated.

Tests (8 new):
- Two name() namespace collision tests (point read and prefix scan
  paths, WKF on/off).
- Inconsistent extract_first via a multi-prefix extractor.
- Concurrent reads against a tree mutated by a background writer/flusher.
- Partitioned filter finish without extractor (locks in notify_key
  always-called invariant).
- Three recovery scenarios: WKF=false same-extractor reopen, legacy
  table + extractor added, extractor change with WKF=true (verifies the
  full-key Bloom fallback via io_skipped_by_filter metric).

No on-disk format changes. All 591 tests pass; cargo clippy clean.
